llama-index-core/llama_index/core/chat_engine/condense_plus_context.py

```diff
@@ -150,41 +183,104 @@ async def _acondense_question(
         chat_history_str = messages_to_history_str(chat_history)
         logger.debug(chat_history_str)
 
-        return await self._llm.apredict(
-            self._condense_prompt_template,
-            question=latest_message,
-            chat_history=chat_history_str,
+        llm_input = self._condense_prompt_template.format(
+            chat_history=chat_history_str, question=latest_message
         )
 
-    def _retrieve_context(self, message: str) -> Tuple[str, List[NodeWithScore]]:
-        """Build context for a message from retriever."""
+        return str(await self._llm.acomplete(llm_input))
+
+    def _get_nodes(self, message: str) -> List[NodeWithScore]:
+        """Generate context information from a message."""
         nodes = self._retriever.retrieve(message)
         for postprocessor in self._node_postprocessors:
             nodes = postprocessor.postprocess_nodes(
                 nodes, query_bundle=QueryBundle(message)
             )
 
-        context_str = "\n\n".join(
-            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
-        )
-        return context_str, nodes
+        return nodes
 
-    async def _aretrieve_context(self, message: str) -> Tuple[str, List[NodeWithScore]]:
-        """Build context for a message from retriever."""
+    async def _aget_nodes(self, message: str) -> List[NodeWithScore]:
+        """Generate context information from a message."""
         nodes = await self._retriever.aretrieve(message)
         for postprocessor in self._node_postprocessors:
             nodes = postprocessor.postprocess_nodes(
                 nodes, query_bundle=QueryBundle(message)
             )
 
-        context_str = "\n\n".join(
-            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
+        return nodes
+
+    def _get_prefix_qa_messages_with_context(
```

Comment:
no way to abstract this implementation + the contextchatengine impl. into one utils function? seems like mostly the same code

Reply:
Hmm maybe, I can give it a shot

Reply:
addressed!